### PR TITLE
Lesson 7 Part 1: Character device driver

### PR DESCRIPTION
**Pull request by Ayrum1158**

### CODE REVIEW

File: 02-KernelModules_CharacterDevice/character_driver.c [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+};
+
+static ssize_t proc_read(struct file *file_p, char __user *usr_buff, size_t len, loff_t *offset)
+{
+	if(*offset != 0)
+		goto end_read;
+
+	char res[256];
+
+	int res_len = sprintf(res, "Buffer info:\n\tallocated size - %d bytes\n\tused size - %d bytes\n", alloc_size, used_size);
```
Comment (oleg-h2o) [inline 02-KernelModules_CharacterDevice/character_driver.c]: A very long line.

File: 02-KernelModules_CharacterDevice/character_driver.c [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+	}
+
+	buffer = kzalloc(alloc_size, GFP_KERNEL);
+
+	alloc_chrdev_region(&dev_num, 0, NUMBER_OF_DEVICES, DEVICE_NAME);
+	chdev_class = class_create(THIS_MODULE, CLASS_NAME);
+	chdev = device_create(chdev_class, NULL, dev_num, NULL, DEVICE_NAME);
+
+	cdev_init(&dev_struct, &fops);
+	cdev_add(&dev_struct, dev_num, NUMBER_OF_DEVICES);
```
Comment (oleg-h2o) [inline 02-KernelModules_CharacterDevice/character_driver.c]: cdev_add() can return an error. Plz handle it.

**Review by oleg-h2o — CHANGES_REQUESTED** [review]
It's almost OK.

File: 02-KernelModules_CharacterDevice/character_driver.c [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+	alloc_chrdev_region(&dev_num, 0, NUMBER_OF_DEVICES, DEVICE_NAME);
+	chdev_class = class_create(THIS_MODULE, CLASS_NAME);
+	chdev = device_create(chdev_class, NULL, dev_num, NULL, DEVICE_NAME);
+
+	int err;
+
+	cdev_init(&dev_struct, &fops);
+	err = cdev_add(&dev_struct, dev_num, NUMBER_OF_DEVICES);
+
```
Comment (oleg-h2o) [inline 02-KernelModules_CharacterDevice/character_driver.c:191]: if (err..) goto cdev_add_err ????

**Review by oleg-h2o — CHANGES_REQUESTED** [review]
Plz add if statement.

File: 02-KernelModules_CharacterDevice/chdev_test.script [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+echo "Unloading module"
+rmmod ${MODULE_NAME}
+
+echo "Trying to load module with buffer size < 1024 (should not work):"
+insmod ${MODULE_FULL_PATH} alloc_size=512
+
+echo "Checking if module is loaded (should be nothing):"
+lsmod | grep "${MODULE_NAME}"
+
+echo "Testing done"
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/chdev_test.script]:
```suggestion
echo "Testing done"

```

Please, add empty line at the end of file

**Review by galkinletter — CHANGES_REQUESTED** [review]
Please fix

**Comment by Ayrum1158** [comment]
Added empty line to the end of the chdev_test.script and "if(err < 0) goto chdev_add_err;" to character_driver.c

**Comment by galkinletter** [comment]
Ok

**Pull request merged**